pkg/algorithm/nodeselect/v1alpha1/select_node.go

```diff
@@ -187,6 +210,116 @@ func (ac *Config) getBlockDevice() (*ndmapis.BlockDeviceList, error) {
 		return nil, err
 	}
-	bdl := bdL.Filter(blockdevice.FilterClaimedDevices, blockdevice.FilterNonInactive)
+	bdl := bdL.Filter(blockdevice.FilterNonInactive)
 	return bdl.BlockDeviceList, nil
 }
+
+//TODO: Make changes in below code befor PR checked in
+// 1) Use Builder Pattern or some approach to present below code
+// 2) Refactor the below code before removing WIP
+
+// ClaimBlockDevice will create BDC for corresponding BD
+func (ac *Config) ClaimBlockDevice(nodeBDs *nodeBlockDevice, spc *apis.StoragePoolClaim) (*NodeClaimedBDDetails, error) {
+	nodeClaimedBDs := &NodeClaimedBDDetails{
+		NodeName:        "",
+		BlockDeviceList: []ClaimedBDDetails{},
+	}
+
+	if nodeBDs == nil || len(nodeBDs.BlockDevices.Items) == 0 {
+		return nil, errors.New("No valid block devices are present to claim")
+	}
+
+	namespace := env.Get(env.OpenEBSNamespace)
+	bdcKubeclient := bdc.NewKubeClient().
+		WithNamespace(namespace)
+	labels := map[string]string{string(apis.StoragePoolClaimCPK): spc.Name}
+
+	nodeClaimedBDs.NodeName = nodeBDs.NodeName
+	pendingBDCCount := 0
+
+	bdcObjList, err := bdcKubeclient.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list block device claim for {%s}", spc.Name)
+	}
+	customBDCObjList := bdc.ListBuilderFromAPIList(bdcObjList)
+
+	for _, bdName := range nodeBDs.BlockDevices.Items {
+		var hostName, bdcName string
+		claimedBD := ClaimedBDDetails{}
+
+		bdObj, err := ac.BlockDeviceClient.Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get block device {%s}", bdName)
+		}
+		hostName = bdObj.Labels[string(apis.HostNameCPK)]
+
+		if bdObj.IsClaimed() {
+			//TODO: Use HasLabel Predicate for below check
+			bdcName = bdObj.Spec.ClaimRef.Name
+			bdcObj := customBDCObjList.GetBlockDeviceClaim(bdcName)
+			if bdcObj == nil {
+				return nil, errors.Errorf("failed to get bdc {%s} for bd {%s}",
+					bdcName, bdName)
+			}
+			if bdcObj.Labels[string(apis.StoragePoolClaimCPK)] != spc.Name {
+				return nil, errors.Errorf("block device %s is already in use", bdName)
+			}
+		} else {
+			bdcName = "bdc-" + string(bdObj.UID)
+			bdcObj := customBDCObjList.GetBlockDeviceClaim(bdcName)
+			if bdcObj != nil {
+				pendingBDCCount++
+				continue
+			}
+			capacity := volume.ByteCount(bdObj.Spec.Capacity.Storage)
+			//TODO: Move below code to some function
+			newBDCObj, err := bdc.NewBuilder().
+				WithName(bdcName).
+				WithNamespace(namespace).
+				WithLabels(labels).
+				WithBlockDeviceName(bdName).
+				WithHostName(hostName).
+				WithCapacity(capacity).
+				WithOwnerReference(spc).
+				Build()
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to build blockd device claim for bd {%s}", bdName)
+			}
+
+			_, err = bdcKubeclient.Create(newBDCObj.Object)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to create block device claim for bdc {%s}", bdcName)
+			}
+
+			isClaimed, err := ac.waitForClaimedStatus(bdName)
```

Comment:
we don't need this

Reply:
No need to check isClaimed.. pendingBDCCount can be just incremented

Reply:
Done